### PR TITLE
ros2 migration

### PR DESCRIPTION
**Pull request by fllay**
I edited witmotion_ros.cpp, witmotion_ros_node.cpp and witmotion.h to add ros2 publisher and service. I still struggle with witmotion_ros_node.cpp.

### CODE REVIEW

**Comment by twdragon** [comment]
@fllay you removed compilation of the underlying library and linking it to the main node from `CMakeLists.txt`. This is why the node does not link properly. Please check your build pipeline.

**Comment by fllay** [comment]
I fixed link problem. Now, only how to use `` in ros2.

**Comment by twdragon** [comment]
@fllay please refer here to solve: https://discourse.ros.org/t/async-executor-in-ros2/1575 

It seems working.

**Comment by twdragon** [comment]
@fllay if you would be successful to test it, I will create the ROS2 branch here, but now I can not unless the code is not feasible to be run.

**Comment by fllay** [comment]
I have tried https://discourse.ros.org/t/async-executor-in-ros2/1575. it seems to be ok but now I have another problem. Somehow, It cannot communicate with serial port.


```
pi@pi-desktop:~/ros2_ws$ ros2 run witmotion witmotion_ros_node --ros-args --params-file /home/pi/ros2_ws/src/witmotion_IMU_ros/config/wt61c.yml
[INFO] [1667615027.287959826] [ROSWitmotionSensorController]: Controller started@@@@@@
Opening device "ttyAMA1" at 115200 baud
Instantiating timer at 30 ms
^C[INFO] [1667615031.384384523] [rclcpp]: signal_handler(signum=2)
[ERROR] [1667615031.387922914] [ROSWitmotionSensorController]: Sensor error: No data acquired during last 3 iterations, please check the baudrate!
[INFO] [1667615031.388039951] [ROSWitmotionSensorController]: Entering SUSPENDED state
QObject::killTimer: Timers cannot be stopped from another thread
QObject::~QObject: Timers cannot be stopped from another thread
QSocketNotifier: Socket notifiers cannot be enabled or disabled from another thread
Suspending TTL connection, please emit RunPoll() again to proceed!
[ERROR] [1667615031.390323922] [ROSWitmotionSensorController]: Sensor error: No data acquired during last 3 iterations, please check the baudrate!
[INFO] [1667615031.390413366] [ROSWitmotionSensorController]: Entering SUSPENDED state
Suspending TTL connection, please emit RunPoll() again to proceed!
[ERROR] [1667615031.390650217] [ROSWitmotionSensorController]: Sensor error: No data acquired during last 3 iterations, please check the baudrate!
[INFO] [1667615031.390727957] [ROSWitmotionSensorController]: Entering SUSPENDED state
Suspending TTL connection, please emit RunPoll() again to proceed!
```

Hardware is same as the one I use with ROS1. I also fixed parameter yaml file so that it can be now used in ROS2. All parameters look ok for me. But the serial port read is fail.

**Comment by twdragon** [comment]
@fllay please try to compile the library separately and run one of the controller applications (I suggest `witmotionctl-jy901`) with the same baudrate on the same port to see what will happen.

**Comment by twdragon** [comment]
@fllay if the controller application will print the result, it should be checked that the `start()` function is called for the sensor controller object to start the Qt event loop with decoder functions, but it seems working already.

**Comment by fllay** [comment]
I think I know where the problem is but I am still looking for a solution. The problem is 

```
    rclcpp::executors::MultiThreadedExecutor executor;
    auto executor_spin_lambda = [&executor]() {
      executor.spin();
    };
    
    executor.add_node(node);
    {
        std::thread spin_thread(executor_spin_lambda);
        spin_thread.join();
    }
   int result = app.exec();
   return result;
```  

After `spin_thread.join();`   this ` int result = app.exec();` code will not run until I hit ctrl-c then I got 

```
Suspending TTL connection, please emit RunPoll() again to proceed!
[ERROR] [1667615031.390323922] [ROSWitmotionSensorController]: Sensor error: No data acquired during last 3 iterations, please check the baudrate!
[INFO] [1667615031.390413366] [ROSWitmotionSensorController]: Entering SUSPENDED state
Suspending TTL connection, please emit RunPoll() again to proceed!
```

then it stops. This output is from ` int result = app.exec();`.  I also try to put `spin_thread.join();` after `int result = app.exec();` but it does not work either. 

The solution is to execute `QCoreApplication app(argc, argv);` in  `std::thread` but I don't know how. Any advice? 

That is the only problem. I have tested ros2 `publisher` using timer. It works just fine.

**Comment by twdragon** [comment]
> I think I know where the problem is but I am still looking for a solution. The problem is
> 
> ```
>     rclcpp::executors::MultiThreadedExecutor executor;
>     auto executor_spin_lambda = [&executor]() {
>       executor.spin();
>     };
>     
>     executor.add_node(node);
>     {
>         std::thread spin_thread(executor_spin_lambda);
>         spin_thread.join();
>     }
>    int result = app.exec();
>    return result;
> ```
> 
> After `spin_thread.join();` this ` int result = app.exec();` code will not run until I hit ctrl-c then I got
> 
> ```
> Suspending TTL connection, please emit RunPoll() again to proceed!
> [ERROR] [1667615031.390323922] [ROSWitmotionSensorController]: Sensor error: No data acquired during last 3 iterations, please check the baudrate!
> [INFO] [1667615031.390413366] [ROSWitmotionSensorController]: Entering SUSPENDED state
> Suspending TTL connection, please emit RunPoll() again to proceed!
> ```
> 
> then it stops. This output is from ` int result = app.exec();`. I also try to put `spin_thread.join();` after `int result = app.exec();` but it does not work either.
> 
> The solution is to execute `QCoreApplication app(argc, argv);` in `std::thread` but I don't know how. Any advice?
> 
> That is the only problem. I have tested ros2 `publisher` using timer. It works just fine.

@fllay here `join()` is the blocking call. I found the following solution [here](https://nicolovaligi.com/articles/concurrency-and-parallelism-in-ros1-and-ros2-application-apis/):
```
// Create a Node and an Executor.
rclcpp::executors::SingleThreadedExecutor executor1;
auto node1 = rclcpp::Node::make_shared("node1");
executor1.add_node(node1);

// Create another.
rclcpp::executors::SingleThreadedExecutor executor2;
auto node2 = rclcpp::Node::make_shared("node2");
executor2.add_node(node2);

// Spin the Executor in a separate thread.
std::thread spinThread([&executor2]() {
    executor2.spin();
});

```

So, please, check is `join()` call could be handled in a different way or to be excluded in favour of another `std::thread` API call.

**Comment by fllay** [comment]
Let take a look at `main()` and its output

```
int main(int argc, char *argv[]) {
  struct sigaction sigIntHandler;
  sigIntHandler.sa_handler = handle_shutdown;
  sigemptyset(&sigIntHandler.sa_mask);
  sigIntHandler.sa_flags = 0;
  sigaction(SIGINT, &sigIntHandler, NULL);

  QCoreApplication app(argc, argv);

  rclcpp::init(argc, argv);

  ROSWitmotionSensorController &controller =
      ROSWitmotionSensorController::Instance();
  auto node = controller.Start();

  rclcpp::executors::SingleThreadedExecutor executor;


  executor.add_node(node);

  

  std::thread spinThread([&executor, &app]() {
    executor.spin();

  });

  spinThread.detach();
  RCLCPP_INFO(rclcpp::get_logger("MinimalPublisher"), "QT spin !!!!!");
  app.exec();


  return 0;
}

```

**Comment by twdragon** [comment]
@fllay well, now we have multithreaded schema working. The `app.exec()` call lives in the main thread, so the timer is stopped incorrectly due to a segmentation fault. The error message you see is issued by [`witmotion::QBaseSerialWitmotionSensorReader`](https://github.com/ElettraSciComp/witmotion_IMU_QT/blob/0169b9e85af02f631bd363b5709014a91d175a93/src/serial.cpp#L14) object instance. In fact, it means that the serial port produces data but the `reader` object does not receive it. This fault is obviously reasoned to [this fragment of code](https://github.com/ElettraSciComp/witmotion_IMU_ros/blob/c518dc5cb1c80d41036af8f4d434e1873ab9b73c/src/witmotion_ros.cpp#L215):
>    ```
>    reader->moveToThread(&reader_thread);
>    connect(&reader_thread, &QThread::finished, reader, &QObject::deleteLater);
>    connect(this, &ROSWitmotionSensorController::RunReader, reader, &QAbstractWitmotionSensorReader::RunPoll);
>    connect(this, &ROSWitmotionSensorController::ConfigureSensor, reader, &QAbstractWitmotionSensorReader::SendConfig);
>    connect(reader, &QAbstractWitmotionSensorReader::Acquired, this, &ROSWitmotionSensorController::Packet);
>    connect(reader, &QAbstractWitmotionSensorReader::Error, this, &ROSWitmotionSensorController::Error);
>    reader_thread.start();
>```

It means that `ReadData()` slot is called at least thrice, so the program is attempting to read the data but no data feasible to decode is available. First, I can advise you to uncomment `qt5_wrap_cpp(MOC_SOURCES include/witmotion_ros.h)` line in `CMakeLists.txt` questioned above. Then: can you please send the parameters of the sensor you used to test it on ROS1 version of the package? I think, there is polling frequency mismatch, due to this the data is produced, but the program reads garbage from the port.

P.S. Did you try to use `witmotionctl-` application in different modes?

**Comment by twdragon** [comment]
@fllay thanks to you! Just let me know when your program passes the test, and then I will create the branch.

**Comment by fllay** [comment]
I fixed segmentation fault when press ctrl-c and edited both config and launch files according to ROS2. I think you can create the branch now. I tested it on wt61C only. I ordered W901c from ali-express but it will take a while to get it.

File: CMakeLists.txt [hunk 1]
```diff
@@ -1,60 +1,48 @@
...
-    tf2
-    sensor_msgs
-    geometry_msgs
-    std_msgs
-    std_srvs
-    nav_msgs
-    tf2_geometry_msgs
-    rosgraph_msgs
-)
+project(witmotion)
```
Comment (twdragon) [inline CMakeLists.txt:2]: @fllay I sent the project already to the ROS documentation generation pipeline, so please rename the project back to `witmotion_ros` not to destroy a naming convention

**Review by twdragon — CHANGES_REQUESTED** [review]
It is needed to change the project name back to `witmotion_ros`, then the updated PR will be merged into `ros2` branch.

**Comment by twdragon** [comment]
@fllay I added the branch, we can now do the name fix, final review and merge.

File: CMakeLists.txt [hunk 2]
```diff
@@ -1,60 +1,48 @@
...
 add_subdirectory(witmotion-uart-qt)
-include_directories(${catkin_INCLUDE_DIRS} include witmotion-uart-qt/include)
 
-find_package(Qt5 REQUIRED COMPONENTS Core SerialPort)
+#set(CMAKE_AUTOMOC ON)
+#qt5_wrap_cpp(MOC_SOURCES include/witmotion_ros.h)
+include_directories(include)
+include_directories(/opt/ros/humble/include/tf2_geometry_msgs)
+include_directories(witmotion-uart-qt/include)
+#add_library(witmotion_ros src/witmotion_ros.cpp include/witmotion_ros.h)
```
[the remark this answers is not in the document]
Comment (fllay) [reply CMakeLists.txt:27]: It is in line 50.

File: CMakeLists.txt [hunk 3]
```diff
@@ -1,60 +1,48 @@
...
+find_package(tf2 REQUIRED)
+find_package(tf2_geometry_msgs REQUIRED)
+find_package(rclcpp REQUIRED)
+find_package(std_srvs REQUIRED)
+find_package(sensor_msgs REQUIRED)
+find_package(geometry_msgs REQUIRED)
+find_package(nav_msgs REQUIRED)
+find_package(rosgraph_msgs REQUIRED)
+find_package(std_msgs REQUIRED)
+find_package(Qt5Core REQUIRED)
```
[the remark this answers is not in the document]
Comment (fllay) [reply CMakeLists.txt:40]: I edit it to be `find_package(Qt5 COMPONENTS SerialPort REQUIRED)` as suggested.

File: CMakeLists.txt (hunk 1 above)
Comment (fllay) [reply CMakeLists.txt:2]: I change project name to `witmotion_ros`.

**Review by twdragon — APPROVED** [review]

**Pull request merged**